qiskit/quantum_info/operators/base_operator.py

```diff
@@ -173,31 +202,57 @@ def reshape(self, input_dims=None, output_dims=None):
             QiskitError: if combined size of all subsystem input dimension or
                          subsystem output dimensions is not constant.
         """
-        clone = copy.copy(self)
+        # Trivial case
+        if num_qubits:
+            if self.num_qubits == num_qubits:
+                return self
+            dim = 2 * (2 ** num_qubits, )
+            if self.dim != dim:
```

Comment:
There might be a bit of a performance gain here for the self.num_qubits defined case here if you did `elif self.num_qubits is not None: raise` before this block but it's probably not worth the extra code complexity

Reply:
I'm not too worried about performance in the paths that lead to an exception anyway so probably not worth extra code complexity like you say